### ISSUE TITLE
samtools mpileup error: "free() malloc() memory corruption"

### ISSUE THREAD

**Issue by zypiner** [samtools/samtools#718]
Hello,

When I executed command 'samtools mpileup -f hg19.fa S12-W.bam', I got the following errors. But strangely there was no errors if I didn`t use -f parameter.
'
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
*** Error in `samtools': free(): invalid next size (fast): 0x00000000012362d0 ***
*** Error in `samtools': malloc(): memory corruption: 0x0000000001204e00 ***
'

I would like to explain this bam file. The reads in this bam file are pair-end reads and are only from two regions, chr4:1806109-1806208 and chr4:1806461-1806561. I extracted these reads from one bam file that was created using 'bwa mem -t 10 xx.r1.fq xx.r2.fq'.

For this bam file, please refer to the attached folder '3_mpileup_NG_merge208_461bam'.

There was another strange item I would like to explain:
When I split this bam file into two bam files by the two region, the mpileup command was executed OK on each separated bam file. But if I merged these two separated bam files into one bam file using 'samtools merge', the mpileup command still can not executed OK.

For the two separated bam files, please refer to the attached folder '1_mpileup_OK_208bam' and '2_mpileup_OK_461bam'. The bam file in the folder '1_mpileup_OK_208bam' is from region chr4:1806109-1806208, and the bam file from the other folder is from region chr4:1806461-1806561.

PS:
1. The version of samtools I used is 'Version: 0.1.19-96b5f2294a', but I had tried the newest version, the same error is still exist.
2. The attached file contains three folders, one folder contains the failed bam, one folder contains bam whose reads are from region chr4:1806109-1806208, the another folder contains bam whose reads are from region chr4:1806461-1806561.

Thank you for your reply.

[bamfiles_3folder.zip](https://github.com/samtools/samtools/files/1283535/bamfiles_3folder.zip)

**Comment by daviesrob** [comment]
Thanks for the bug report and the test case.

The problem is caused by some of your alignments, where the query length from the CIGAR string is longer than the sequence in the read.  For example:

```
$ samtools view 3_mpileup_NG_merge208_461bam/S12-W.bam | grep ST-E00142:333:H32V7ALXX:2:2111:14763:59411
ST-E00142:333:H32V7ALXX:2:2111:14763:59411	163	chr4	1806461	60	20M1I81M	=	1806461	101	CCCCCCCTTCTGTCCCAGCACTGACCCCCGGCGGAACCTCCACCCCCTGTCCCCCCCGCGGCGCCAACCTGCCCCTGCTGACCCA	<F<-<A-<--A<-A7F-7A--<7<AJ-<AJ---7--7-77<-7--A------AJF-7-7--7FJJ<7J-F-7-AJA<F-7-JJ77	NM:i:6	MD:Z:12G19T1T8G7G3A2929AS:Z:55	XS:Z:15
ST-E00142:333:H32V7ALXX:2:2111:14763:59411	83	chr4	1806461	60	20M1I81M	=	1806461	-101	CCCCCCCTTCTGGCCCAGCACTGACCCCCGGCTGTCCCTCCACGCCCTGTCGCCCACGCGGCGCCAACCTGCCCCTGCTGACCCAAGCAGGTGTCCCTGGA	JJJJJJJFJJJJJFJJJJFJJJJ-JFAAJJJJJF7-<AJAAAJFJJFJJFJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJFJJJJJJJJJJJJFJJJJJAF	NM:i:1	MD:Z:35A65	AS:Z:92	XS:Z:13
```

In the both of these reads, the CIGAR string indicates 20+1+81 = 102 bases should be present, but the sequences only have 85 bases in the first read and 101 in the second.  Quite a few of the other reads show a similar problem.  The bad reads are in both the 2_mpileup_OK_461bam and 3_mpileup_NG_merge208_461bam versions of the file, and in fact I can reproduce the crash in both.  I'm afraid I don't know why it worked for you in 2_mpileup_OK_461bam, but as the results of this sort of memory corruption can be fairly random, I suspect you may have just got away with it that time.

As your bam file is corrupt, you need to go back to your original bwa mem output to see if that's OK, and then work forward though any other processing you did to find out where the error occurred.

I think the solution for this will be to make the BAM reader check that the CIGAR length is consistent.  Both the SAM and CRAM readers already do this, so if you use either of these as the input format then you get a sensible message about what the problem is:

```
$ samtools view -h /tmp/i718/3_mpileup_NG_merge208_461bam/S12-W.bam | samtools mpileup -o /dev/null -f hgc19.fa -
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
[E::sam_parse1] CIGAR and query sequence are of different length
[W::sam_read1] Parse error at line 8006
```

**Comment by zypiner** [comment]
Thank you for your timely reply. It is really helpful for me.

Actually, I  had modified the bam file manually after executing bwa mem. Because the start part of the read are  primer sequence(maybe 20~40bp) and I want to delete the primer sequence, I deleted the primer sequence of the read and modified cigar, mapping position, MD and so on that were influenced by deleting part sequence of the read. I thought that it was my process that  led to the strange reads.

By using the newest samtools I could check whether CIGAR length is consistent. It is so useful for me to debug my program. Thank you again.

**Comment by daviesrob** [comment]
This has been fixed by HTSlib commit [1cc6afd1](https://github.com/samtools/htslib/commit/1cc6afd1).  It adds the same check on CIGAR integrity to BAM as was already done for SAM and CRAM.  This means on your test data, you now get a message like this:
```
[E::bam_read1] CIGAR and query sequence lengths differ for ST-E00142:333:H32V7ALXX:2:2111:14763:59411
```
Hopefully this should make similar problems easier to solve.
